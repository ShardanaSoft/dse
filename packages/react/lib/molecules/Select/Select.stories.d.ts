/// <reference types="react" />
declare const _default: {
    title: string;
    decorators: import("@storybook/csf").DecoratorFunction<import("@storybook/csf").AnyFramework, import("@storybook/csf").Args>[];
};
export default _default;
export declare const Common: () => JSX.Element;
export declare const RenderOption: () => JSX.Element;
export declare const CustomLabel: () => JSX.Element;
