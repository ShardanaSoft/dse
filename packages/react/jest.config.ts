import type { Config } from "jest";

const config: Config = {
  roots: ["<rootDir>/src"],
  testRegex: "(/.*\\.test)\\.(ts|tsx)$",
  setupFilesAfterEnv: ["<rootDir>/jest.setup.ts"],
  moduleNameMapper: {
    "\\.(css|less|sass|scss)$": "<rootDir>/__mocks__/styleMock.js",
  },
  testEnvironment: "jsdom",
};

export default config;

