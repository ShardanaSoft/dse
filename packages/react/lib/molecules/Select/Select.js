import React, { useState, useRef, useEffect, createRef } from 'react';
import Text from '../../atoms/Text/Text.js';
import '@ds.e/scss/lib/Select.css';

const getNextOptionIndex = (currentIndex, options) => {
    if (currentIndex === null)
        return 0;
    if (currentIndex === options.length - 1)
        return 0;
    return currentIndex + 1;
};
const getPreviousOptionIndex = (currentIndex, options) => {
    if (currentIndex === null)
        return 0;
    if (currentIndex === 0)
        return options.length - 1;
    return currentIndex - 1;
};
const Select = ({ onOptionSelected: handler, label = "Select an option", options = [], renderOption, }) => {
    const [isOpen, setIsOpen] = useState(false);
    const [selectedIndex, setSelectedIndex] = useState(null);
    const [highlightedIndex, setHighlightedIndex] = useState(null);
    const labelRef = useRef(null);
    const [optionRefs, setOptionRefs] = useState([]);
    const [overlayTop, setOverlayTop] = useState(0);
    const onOptionSelected = (option, optionIndex) => {
        if (handler)
            handler(option, optionIndex);
        setSelectedIndex(optionIndex);
        setIsOpen((prev) => !prev);
    };
    const onLabelClick = () => {
        if (!isOpen) {
            highlightOption(0);
        }
        setIsOpen((prev) => !prev);
    };
    useEffect(() => {
        setOverlayTop(labelRef.current?.offsetHeight || 0);
    }, [labelRef.current?.offsetHeight]);
    const KEY_CODES = {
        ENTER: "Enter",
        SPACE: " ",
        DOWN_ARROW: "ArrowDown",
        UP_ARROW: "ArrowUp",
        ESC: "Escape",
    };
    const highlightOption = (optionIndex) => {
        setHighlightedIndex(optionIndex);
    };
    const onButtonKeyDown = (event) => {
        event.preventDefault();
        console.log("button keydown", event.key);
        optionRefs[0].current?.focus();
        if ([KEY_CODES.DOWN_ARROW, KEY_CODES.ENTER, KEY_CODES.SPACE].includes(event.key)) {
            if (!isOpen) {
                setIsOpen(true);
                highlightOption(0);
            }
        }
    };
    useEffect(() => {
        setOptionRefs(options.map((_) => createRef()));
    }, [options.length]);
    useEffect(() => {
        if (highlightedIndex !== null && isOpen) {
            const ref = optionRefs[highlightedIndex];
            if (ref && ref.current)
                ref.current.focus();
        }
    }, [isOpen, highlightedIndex]);
    const onOptionKeyDown = (event) => {
        switch (event.key) {
            case KEY_CODES.ESC:
                setIsOpen(false);
                break;
            case KEY_CODES.DOWN_ARROW:
                highlightOption(getNextOptionIndex(highlightedIndex, options));
                break;
            case KEY_CODES.UP_ARROW:
                highlightOption(getPreviousOptionIndex(highlightedIndex, options));
                break;
            case KEY_CODES.ENTER:
                onOptionSelected(options[highlightedIndex], highlightedIndex);
                break;
        }
    };
    return (React.createElement("div", { className: "dse-select" },
        React.createElement("button", { "data-testid": "DseSelectButton", onKeyDown: onButtonKeyDown, "aria-controls": "dse-select-list", "aria-haspopup": true, "aria-expanded": isOpen ? true : undefined, ref: labelRef, className: "dse-select__label", onClick: onLabelClick },
            React.createElement(Text, null, selectedIndex === null ? label : options[selectedIndex].label),
            React.createElement("svg", { className: `dse-select__caret ${isOpen ? "dse-select__caret--open" : "dse-select__caret--closed"}`, width: "1rem", height: "1rem", fill: "none", strokeLinecap: "round", strokeLinejoin: "round", strokeWidth: 2, viewBox: "0 0 24 24", stroke: "currentColor" },
                React.createElement("path", { d: "M19 9l-7 7-7-7" }))),
        React.createElement("ul", { role: "menu", id: "dse-select-list", style: { top: overlayTop }, className: `dse-select__overlay ${isOpen ? "dse-select__overlay--open" : ""}` }, options.map((option, optionIndex) => {
            const isSelected = selectedIndex === optionIndex;
            const isHighlighted = highlightedIndex === optionIndex;
            const ref = optionRefs[optionIndex];
            const renderOptionProps = {
                option,
                isSelected,
                getOptionRecommendedProps: (overrideProps = {}) => {
                    return {
                        ref,
                        role: "menuitemradio",
                        "aria-label": option.label,
                        "aria-checked": isSelected ? true : undefined,
                        onKeyDown: onOptionKeyDown,
                        tabIndex: isHighlighted ? -1 : 0,
                        onMouseEnter: () => highlightOption(optionIndex),
                        onMouseLeave: () => highlightOption(null),
                        className: `dse-select__option ${isSelected && "dse-select__option--selected"} ${isHighlighted && "dse-select__option--highlighted"}`,
                        onClick: () => onOptionSelected(option, optionIndex),
                        key: option.value,
                        ...overrideProps,
                    };
                },
            };
            if (renderOption)
                return renderOption(renderOptionProps);
            return (React.createElement("li", { ...renderOptionProps.getOptionRecommendedProps() },
                React.createElement(Text, null, option.label),
                isSelected && (React.createElement("svg", { width: "1rem", xmlns: "http://www.w3.org/2000/svg", fill: "none", viewBox: "0 0 24 24", strokeWidth: 1.5, stroke: "currentColor" },
                    React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M4.5 12.75l6 6 9-13.5" })))));
        }))));
}; //test comments for git

export { Select as default };
//# sourceMappingURL=Select.js.map
