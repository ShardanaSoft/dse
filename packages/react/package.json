{
  "name": "@navoxds.e/react",
  "version": "0.5.0",
  "description": "",
  "main": "lib/index.js",
  "type": "module",
  "scripts": {
    "build": "rollup -c",
    "dev": "npm run build -- --watch",
    "test": "jest --verbose",
    "test:watch": "npm run test -- --watch",
    "start-storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook"
  },
  "author": "",
  "license": "ISC",
  "files": [
    "lib"
  ],
  "devDependencies": {
    "@babel/core": "^7.20.12",
    "@babel/preset-env": "^7.20.2",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@storybook/addon-a11y": "^6.5.16",
    "@storybook/addon-actions": "^6.5.15",
    "@storybook/addon-essentials": "^6.5.15",
    "@storybook/addon-interactions": "^6.5.15",
    "@storybook/addon-knobs": "^6.4.0",
    "@storybook/addon-links": "^6.5.15",
    "@storybook/builder-webpack4": "^6.5.15",
    "@storybook/manager-webpack4": "^6.5.15",
    "@storybook/react": "^6.5.15",
    "@storybook/testing-library": "^0.0.13",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@types/jest": "^29.2.5",
    "@types/react": "^18.0.26",
    "babel-loader": "^8.3.0",
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.3.1",
    "react": "^18.2.0",
    "rollup": "^3.5.1",
    "rollup-plugin-typescript2": "^0.34.1",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.3"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "cfcfe7cfdc7f30600f26ea985efd2ff9fc2c0713"
}
