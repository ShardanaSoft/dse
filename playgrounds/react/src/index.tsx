import ReactDOMClient from "react-dom/client";
import React from "react";
import { Color, Margin, Select, Text } from "@navoxds.e/react";

const options = [
  {
    label: "Sweet pink",
    value: "pink",
  },
  {
    label: "bitter green",
    value: "green",
  },
  {
    label: "sad dark",
    value: "dark",
  },
];

const container = document.getElementById("root");
const root = ReactDOMClient.createRoot(container!);

root.render(
  <>
    <Color hexCode="#18723add" width="xxl" height="lg" />
    <Margin>
      <Text size="sm">Text test</Text>
    </Margin>
    <Select
      options={options}
      label="Please select "
      // renderOption={({ option, getOptionRecommendedProps }) => (
      //   <p {...getOptionRecommendedProps()}>{option.label}</p>
      // )}
    />
  </>
);

