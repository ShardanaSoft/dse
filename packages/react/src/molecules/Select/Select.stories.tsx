import React from "react";
import Select from "./Select";

import { withA11y } from "@storybook/addon-a11y";

export default {
  title: "Molecules/Select",
  decorators: [withA11y],
};

const options = [
  {
    label: "Sweet pink",
    value: "pink",
  },
  {
    label: "bitter green",
    value: "green",
  },
  {
    label: "sad dark",
    value: "dark",
  },
];
export const Common = () => <Select options={options} />;

export const RenderOption = () => (
  <Select
    options={options}
    renderOption={({ getOptionRecommendedProps, option, isSelected }) => (
      <span {...getOptionRecommendedProps()}>
        {option.value} | {isSelected && "IS SELECTED"}
      </span>
    )}
  />
);

export const CustomLabel = () => (
  <Select label="Select a color pls" options={options} />
);

