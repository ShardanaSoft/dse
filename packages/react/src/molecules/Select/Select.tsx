import React, {
  createRef,
  KeyboardEventHandler,
  useEffect,
  useRef,
  useState,
} from "react";

import Text from "../../atoms/Text";
import "@ds.e/scss/lib/Select.css";
interface SelectOption {
  label: string;
  value: string;
}
interface RenderOptionProps {
  isSelected: boolean;
  option: SelectOption;
  getOptionRecommendedProps: (overrideProps?: Object) => Object;
}

interface SelectProps {
  label?: string;
  options?: SelectOption[];
  onOptionSelected?: (option: SelectOption, optionIndex: number) => void;
  renderOption?: (props: RenderOptionProps) => React.ReactNode;
}

const getNextOptionIndex = (
  currentIndex: number | null,
  options: Array<SelectOption>
) => {
  if (currentIndex === null) return 0;

  if (currentIndex === options.length - 1) return 0;

  return currentIndex + 1;
};

const getPreviousOptionIndex = (
  currentIndex: number | null,
  options: Array<SelectOption>
) => {
  if (currentIndex === null) return 0;

  if (currentIndex === 0) return options.length - 1;

  return currentIndex - 1;
};

const Select: React.FC<SelectProps> = ({
  onOptionSelected: handler,
  label = "Select an option",
  options = [],
  renderOption,
}) => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [selectedIndex, setSelectedIndex] = useState<number | null>(null);
  const [highlightedIndex, setHighlightedIndex] = useState<number | null>(null);
  const labelRef = useRef<HTMLButtonElement>(null);
  const [optionRefs, setOptionRefs] = useState<
    React.RefObject<HTMLLIElement>[]
  >([]);
  const [overlayTop, setOverlayTop] = useState<number>(0);

  const onOptionSelected = (option: SelectOption, optionIndex: number) => {
    if (handler) handler(option, optionIndex);

    setSelectedIndex(optionIndex);
    setIsOpen((prev) => !prev);
  };

  const onLabelClick = () => {
    if (!isOpen) {
      highlightOption(0);
    }
    setIsOpen((prev) => !prev);
  };

  useEffect(() => {
    setOverlayTop(labelRef.current?.offsetHeight || 0);
  }, [labelRef.current?.offsetHeight]);

  const KEY_CODES = {
    ENTER: "Enter",
    SPACE: " ",
    DOWN_ARROW: "ArrowDown",
    UP_ARROW: "ArrowUp",
    ESC: "Escape",
  };

  const highlightOption = (optionIndex: number | null) => {
    setHighlightedIndex(optionIndex);
  };

  const onButtonKeyDown: KeyboardEventHandler = (event) => {
    event.preventDefault();

    console.log("button keydown", event.key);
    optionRefs[0].current?.focus();
    if (
      [KEY_CODES.DOWN_ARROW, KEY_CODES.ENTER, KEY_CODES.SPACE].includes(
        event.key
      )
    ) {
      if (!isOpen) {
        setIsOpen(true);

        highlightOption(0);
      }
    }
  };

  useEffect(() => {
    setOptionRefs(options.map((_) => createRef<HTMLLIElement>()));
  }, [options.length]);

  useEffect(() => {
    if (highlightedIndex !== null && isOpen) {
      const ref = optionRefs[highlightedIndex];

      if (ref && ref.current) ref.current.focus();
    }
  }, [isOpen, highlightedIndex]);

  const onOptionKeyDown: KeyboardEventHandler = (event) => {
    switch (event.key) {
      case KEY_CODES.ESC:
        setIsOpen(false);
        break;
      case KEY_CODES.DOWN_ARROW:
        highlightOption(getNextOptionIndex(highlightedIndex, options));
        break;

      case KEY_CODES.UP_ARROW:
        highlightOption(getPreviousOptionIndex(highlightedIndex, options));
        break;

      case KEY_CODES.ENTER:
        onOptionSelected(options[highlightedIndex!], highlightedIndex!);
        break;
    }
  };

  return (
    <div className="dse-select">
      <button
        data-testid="DseSelectButton"
        onKeyDown={onButtonKeyDown}
        aria-controls="dse-select-list"
        aria-haspopup={true}
        aria-expanded={isOpen ? true : undefined}
        ref={labelRef}
        className="dse-select__label"
        onClick={onLabelClick}
      >
        <Text>
          {selectedIndex === null ? label : options[selectedIndex].label}
        </Text>
        <svg
          className={`dse-select__caret ${
            isOpen ? "dse-select__caret--open" : "dse-select__caret--closed"
          }`}
          width="1rem"
          height="1rem"
          fill="none"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          stroke="currentColor"
        >
          <path d="M19 9l-7 7-7-7" />
        </svg>
      </button>
      {
        <ul
          role={"menu"}
          id="dse-select-list"
          style={{ top: overlayTop }}
          className={`dse-select__overlay ${
            isOpen ? "dse-select__overlay--open" : ""
          }`}
        >
          {options.map((option, optionIndex) => {
            const isSelected = selectedIndex === optionIndex;
            const isHighlighted = highlightedIndex === optionIndex;

            const ref = optionRefs[optionIndex];

            const renderOptionProps: RenderOptionProps = {
              option,
              isSelected,
              getOptionRecommendedProps: (overrideProps = {}) => {
                return {
                  ref,
                  role: "menuitemradio",
                  "aria-label": option.label,
                  "aria-checked": isSelected ? true : undefined,
                  onKeyDown: onOptionKeyDown,
                  tabIndex: isHighlighted ? -1 : 0,
                  onMouseEnter: () => highlightOption(optionIndex),
                  onMouseLeave: () => highlightOption(null),
                  className: `dse-select__option ${
                    isSelected && "dse-select__option--selected"
                  } ${isHighlighted && "dse-select__option--highlighted"}`,
                  onClick: () => onOptionSelected(option, optionIndex),
                  key: option.value,

                  ...overrideProps,
                };
              },
            };

            if (renderOption) return renderOption(renderOptionProps);

            return (
              <li {...renderOptionProps.getOptionRecommendedProps()}>
                <Text>{option.label}</Text>
                {isSelected && (
                  <svg
                    width={"1rem"}
                    xmlns="http://www.w3.org/2000/svg"
                    fill="none"
                    viewBox="0 0 24 24"
                    strokeWidth={1.5}
                    stroke="currentColor"
                  >
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      d="M4.5 12.75l6 6 9-13.5"
                    />
                  </svg>
                )}
              </li>
            );
          })}
        </ul>
      }
    </div>
  );
}; //test comments for git

export default Select;

