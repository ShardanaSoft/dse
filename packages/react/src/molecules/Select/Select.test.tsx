import React from "react";

import Select from "./Select";

import { render, fireEvent } from "@testing-library/react";

const options = [
  {
    label: "Sweet pink",
    value: "pink",
  },
  {
    label: "bitter green",
    value: "green",
  },
  {
    label: "sad dark",
    value: "dark",
  },
];
it("renders all options added to it", () => {
  const { getAllByRole, getByTestId } = render(<Select options={options} />);

  const button = getByTestId("DseSelectButton");
  fireEvent.click(button);

  expect(getAllByRole("menuitemradio")).toHaveLength(options.length);
});

it("renders options using custom renderOption method if passed as prop", () => {
  const { getAllByTestId, getByTestId } = render(
    <Select
      options={options}
      renderOption={({ option, getOptionRecommendedProps }) => (
        <li {...getOptionRecommendedProps()} data-testid="CustomRenderOption">
          {option.label}
        </li>
      )}
    />
  );

  const button = getByTestId("DseSelectButton");
  fireEvent.click(button);

  const customLi = getAllByTestId("CustomRenderOption");

  expect(customLi).toHaveLength(options.length);
});

it("calls the onOptionSelected prop with selected option and its index if passed", () => {
  const onOptionSelectedMock = jest.fn();
  const { getByTestId, getAllByRole } = render(
    <Select options={options} onOptionSelected={onOptionSelectedMock} />
  );

  const button = getByTestId("DseSelectButton");
  fireEvent.click(button);

  const [firstListItem] = getAllByRole("menuitemradio");

  fireEvent.click(firstListItem);

  expect(onOptionSelectedMock).toHaveBeenCalledWith(options[0], 0);
});

it("the button label changes to the selected option label", () => {
  const { getByTestId, getAllByRole } = render(<Select options={options} />);

  const button = getByTestId("DseSelectButton");
  fireEvent.click(button);

  const [firstListItem] = getAllByRole("menuitemradio");

  fireEvent.click(firstListItem);

  expect(button).toHaveTextContent(options[0].label);
});

it("snapshot of the selected option state", () => {
  const { getByTestId, getAllByRole, asFragment } = render(
    <Select options={options} />
  );

  const button = getByTestId("DseSelectButton");
  fireEvent.click(button);

  const [firstListItem] = getAllByRole("menuitemradio");
  fireEvent.click(firstListItem);

  expect(asFragment()).toMatchSnapshot();
});

it("snapshot of the base state", () => {
  const { asFragment } = render(<Select options={options} />);

  expect(asFragment()).toMatchSnapshot();
});

it("snapshot of the options menu open state", () => {
  const { asFragment, getByTestId } = render(<Select options={options} />);

  const button = getByTestId("DseSelectButton");
  fireEvent.click(button);

  expect(asFragment()).toMatchSnapshot();
});

it("can customize select label", () => {
  const customLabel = "myCustomLabel";
  const { getByTestId } = render(
    <Select options={options} label={customLabel} />
  );

  const button = getByTestId("DseSelectButton");

  expect(button).toHaveTextContent(customLabel);
});

