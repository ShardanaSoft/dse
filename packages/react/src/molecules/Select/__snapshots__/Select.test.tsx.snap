// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Select an option
      </p>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
      >
        <path
          d="M19 9l-7 7-7-7"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the options menu open state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-expanded="true"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Select an option
      </p>
      <svg
        class="dse-select__caret dse-select__caret--open"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
      >
        <path
          d="M19 9l-7 7-7-7"
        />
      </svg>
    </button>
    <ul
      class="dse-select__overlay"
      id="dse-select-list"
      role="menu"
      style="top: 0px;"
    >
      <li
        aria-label="Sweet pink"
        class="dse-select__option false dse-select__option--highlighted"
        role="menuitemradio"
        tabindex="-1"
      >
        <p
          class="dse-text dse-text-base"
        >
          Sweet pinkOK
        </p>
      </li>
      <li
        aria-label="bitter green"
        class="dse-select__option false false"
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          bitter greenOK
        </p>
      </li>
      <li
        aria-label="sad dark"
        class="dse-select__option false false"
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          sad darkOK
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the selected option state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Sweet pink
      </p>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
      >
        <path
          d="M19 9l-7 7-7-7"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
